src/cas/utils/validation_utils.py

```diff
@@ -72,3 +84,263 @@ def validate_labelset_markers(annotation: Dict[str, Any], marker_list: List[str]
+                    f"{cas_inferred_hierarchy[child]} relaation from cas."
+                )
+
+    # Report mismatches
+    if mismatches:
+        for issue in mismatches:
+            logging.warning(issue)
+        return False
+    else:
+        logging.info("Parent-child relationships are consistent between CAS and OBS.")
+        return True
+
+
+def infer_obs_cell_hierarchy(
+    obs: Union[pd.DataFrame, CapAnnDataDF], cas_ranks: Dict[str, int]
+) -> Tuple[Dict[Any, Optional[Any]], list[str]]:
+    """
+    Infers a direct parent-child hierarchy between cell labels based on row co-occurrence.
+
+    This function analyzes the hierarchical relationships between cell labels by comparing their
+    row indices in the `obs` DataFrame. A label is considered a child if its row indices are fully
+    contained within another label's indices. The closest (direct) parent is selected based on
+    CAS-defined ranks.
+
+    Parameters:
+        obs (pd.DataFrame): The AnnData `obs` DataFrame containing labelset columns.
+        cas_ranks (Dict[str, int]): A dictionary mapping cell labels to their CAS-defined rank,
+                                    where lower values indicate higher ranks.
+
+    Returns:
+        Tuple:
+            - Dict[Any, Optional[Any]]: A dictionary mapping each cell label to its inferred direct parent.
+                                        Labels without a parent are assigned `None`.
+            - list[str]: A list of warnings for labels that have identical row memberships, indicating
+                         potential duplicate labelsets.
+    """
+    labelset_list = [k for k, v in cas_ranks.items()]
+
+    # Compute row indices for each cell label within each labelset
+    cell_label_row_indices = {
+        labelset: {
+            label: set(obs.index[obs[labelset] == label])
+            for label in obs[labelset].dropna().unique()
+        }
+        for labelset in labelset_list
+    }
+
+    child_parent_map = {}
+    identical_labelset_warning = []
+
+    # Infer direct parent-child relationships based on row containment
+    for labelset_child in labelset_list:
+        for child_label, child_rows in cell_label_row_indices[labelset_child].items():
+            possible_parents = {}
+
+            for labelset_parent in labelset_list:
+                for parent_label, parent_rows in cell_label_row_indices[
+                    labelset_parent
+                ].items():
+                    if child_label == parent_label:  # Prevent self-matching
+                        continue
+                    if child_rows.issubset(parent_rows) and cas_ranks.get(
+                        labelset_parent, float("-inf")
+                    ) > cas_ranks.get(
+                        labelset_child, float("-inf")
+                    ):  # Parent fully contains child
+                        possible_parents[parent_label] = cas_ranks.get(
+                            labelset_parent, float("-inf")
```

Comment:
Couldn't understand why we need `float("-inf")`. `labelset_parent` is member of `labelset_list` which is generated from `cas_ranks`. So, every `labelset_parent` should exist in the `cas_ranks`. We can remove labelsets without rank from `cas_ranks` though.